### PR TITLE
[query] Fix MakeNDArray prune rules

### PR DESCRIPTION
**Pull request by tpoterba**
Also add more typechecking throughout. Might need to disable that.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Pretty.scala [hunk 1]
```diff
@@ -252,7 +252,7 @@ object Pretty {
             case AggLet(name, _, _, isScan) => prettyIdentifier(name) + " " + prettyBooleanLiteral(isScan)
             case TailLoop(name, args, _) => prettyIdentifier(name) + " " + prettyIdentifiers(args.map(_._1).toFastIndexedSeq)
             case Recur(name, _, t) => prettyIdentifier(name) + " " + t.parsableString()
-            case Ref(name, _) => prettyIdentifier(name)
+            case Ref(name, _) => prettyIdentifier(name) + " " + ir.typ.toString
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/Pretty.scala:255]: Let's back this out for legibility in the logs?

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 2]
```diff
@@ -1429,6 +1429,8 @@ case class TableMapRows(child: TableIR, newRow: IR) extends TableIR {
   override def partitionCounts: Option[IndexedSeq[Long]] = child.partitionCounts
 
   protected[ir] override def execute(ctx: ExecuteContext): TableValue = {
+    val e = Env[Type]("row" -> child.typ.rowType, "global" -> child.typ.globalType)
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala:1432]: I think we want to back this out in favor of the typechecks you've already added in lowering passes.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LoweringPipeline.scala [hunk 3]
```diff
@@ -12,12 +12,16 @@ case class LoweringPipeline(lowerings: IndexedSeq[LoweringPass]) {
...
+    TypeCheck.withContext(x, "initial check")
+
+    if (optimize) {
       x = Optimize(x, noisy = true, context = "Lowerer, initial IR", ctx)
+    }
 
     lowerings.foreach { l =>
       try {
         x = l.apply(ctx, x)
+        TypeCheck.withContext(x, l.context)
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LoweringPipeline.scala:24]: Do we want to do this once for each pass, after the optional Optimize? Doesn't give us the granularity we necessarily want, but I'm thinking maybe for non-testing purposes the most important typecheck is right before execution?

File: hail/src/main/scala/is/hail/expr/ir/Pretty.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Pretty.scala:255]: oh, this is invalid anyway

**Comment by tpoterba** [comment]
I'm gonna back out all of the typechecking stuff and introduce that in a separate PR

**Pull request merged**